### ISSUE TITLE
docker exec --privileged are not work properly as it was before 1.10

### ISSUE THREAD

**Issue by coolljt0725** [docker/docker#21984]
**Output of `docker version`:**

```
Client:
 Version:      1.11.0-dev
 API version:  1.24
 Go version:   go1.5.3
 Git commit:   348d902
 Built:        Wed Apr 13 08:49:37 2016
 OS/Arch:      linux/amd64
 Experimental: true

Server:
 Version:      1.11.0-dev
 API version:  1.24
 Go version:   go1.5.3
 Git commit:   348d902
 Built:        Wed Apr 13 08:49:37 2016
 OS/Arch:      linux/amd64
 Experimental: true

```

**Output of `docker info`:**

```
Containers: 183
 Running: 0
 Paused: 0
 Stopped: 183
Images: 8
Server Version: 1.11.0-dev
Storage Driver: overlay
 Backing Filesystem: xfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge
Kernel Version: 4.5.0
Operating System: Fedora 22 (Twenty Two)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 3.623 GiB
Name: fedora
ID: SR3L:JLSW:WFJM:TYGJ:INKO:JHKO:6FZJ:ZJI5:XI5A:QX4K:GC7X:SIYZ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 13
 Goroutines: 38
 System Time: 2016-04-13T04:52:54.046300698-04:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure registries:
 127.0.0.0/8

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Steps to reproduce the issue:**
1. docker run -tid --name foo busybox sh
2. docker exec -ti --privileged foo sh -c "mkdir mnt && mount -t tmpfs none /mnt"
```
[lei@fedora ~]$ docker run -tid --name foo busybox
46156f85d76d7b23ed38e77607be5615ca7b14cf861d5c2f2b5cff2ed8e26b63
[lei@fedora ~]$ docker exec -ti --privileged foo sh -c "mkdir mnt && mount -t tmpfs none /mnt"
mount: permission denied (are you root?)
```

**Describe the results you received:**

`mount: permission denied (are you root?)`
**Describe the results you expected:**

it should mount successfully.
**Additional information you deem important (e.g. issue happens only occasionally):**
This also doesn't work on version 1.10
The reason is that the container start with default seccomp file which deny the mount syscall
and the exec use the same seccomp file as the container stared.

Should we use unconfined seccomp rule if docker exec with --privileged?

**Comment by tonistiigi** [comment]
I agree. This isn't consistent with apparmor profile that is disabled when you use `--privileged` and do not specify a custom profile.

**Comment by justincormack** [comment]
Ah yes, `docker run --privileged` disables seccomp, but it looks like `docker exec` does not. There are no options to set security profiles on exec so you cannot override, so I guess this is incorrect behaviour. I can look at this.

**Comment by thaJeztah** [comment]
Thanks!

**Comment by jfrazelle** [comment]
Whoa I'm surprised it took this long to catch, (makes note in brain not
many use this feature)

On Thursday, April 14, 2016, Sebastiaan van Stijn <notifications@github.com>
wrote:

> Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/issues/21984#issuecomment-209875414>
>


-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>

**Comment by justincormack** [comment]
So it is a bit more complicated than I thought. `seccomp` is not part of the per process profile, but the host config. What this means is that if you run the initial process as `privileged`, or set a seccomp profile on it, then that applies to any use of `docker exec`, so they become similarly privileged. So:

```
docker run -itd --privileged alpine sh
4bebbcb57c887f03163c3f271d075a3313ca9a04abda7e3bc00c8a6ee1368021
whale:daemon justin$ docker exec 4bebbcb57c887f03163c3f271d075a3313ca9a04abda7e3bc00c8a6ee1368021 mount -t tmpfs none /tmp
```

Actually works. The `--privileged` flag on exec does affect the per process things, for example capabilities.

So changing this involves changing the OCI spec I think, will investigate further.

**Comment by justincormack** [comment]
There is an ongoing discussion about `exec` in the OCI spec, eg in https://github.com/opencontainers/runtime-spec/issues/345

Also what should be per process: https://groups.google.com/a/opencontainers.org/forum/#!msg/dev/Te49anfvghc/5nys-ntoCgAJ

These need to change before we can resolve this issue.

**Comment by rgabo** [comment]
We're trying to limit capabilities to avoid `--privileged` and arrived at `--cap-add SYS_ADMIN --cap-add MKNOD --device /dev/fuse --security-opt seccomp=unconfined` (or a non-default seccomp profile) to use `s3fs` in the container.

Unfortunately `docker exec` does not take any of the security arguments so we have to revert to `--privileged` for the time being.

Maybe `--cap-add` and friends can make their way to `docker exec` more easily?

**Comment by justincormack** [comment]
You can, as a workaround until we sort this out, do a privileged `run` to add a mount inside an exitsing container that is not privileged:

`docker run --privileged --pid=host debian nsenter -t $(docker inspect -f '{{.State.Pid}}' 42c91f54478a) -m mount -t tmpfs tmpfs /tmp`

where `42c91f54478a` is the container id of the running container. That enters the mount namespace of the container to add the additional mount, but has to do it indirectly as there is currently no way to enter a mount namespace directly other than via `exec`.

**Comment by rgabo** [comment]
That's a neat trick and thanks for the amazing 12 minute response time :wink:

**Comment by brendangregg** [comment]
I think this also breaks how we were using Linux perf for container profiling (#32018).

**Comment by jvican** [comment]
I'm running into this as well. Is there any way I can enable the previous behaviour without touching the application running docker? Any kind of global configuration?

**Comment by jelischer** [comment]
Still an issue in 2021.  

+ sudo sh -c 'echo 0 > /sys/class/net/testif0/bridge/multicast_snooping'
sh: /sys/class/net/testif0/bridge/multicast_snooping: Read-only file system

works if you create the container with --privileged but not if you use exec.

**Comment by derickson2402** [comment]
I found an interesting issue.

I was trying to get ```perf``` to work in a ```centos:8``` based container, and along the way I found that specifiying ```--privileged --cap-add SYS_ADMIN``` will not allow me to use ```perf```, somehow adding ```--cap-add SYS_ADMIN``` downgrades the permissions for it to work (or at least for me to fix ```/proc/sys/kernel/kptr_restrict```)
